File: generated-code/fsharp/codecraft/FileReadWrite/Vec2Single.fs
#nowarn "0058"

namespace TransGenTest

/// 2 dimensional vector.
type Vec2Single = {
    /// `x` coordinate of the vector
    X: single;
    /// `y` coordinate of the vector
    Y: single;
} with

    /// Write Vec2Single to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write this.X
        writer.Write this.Y

    /// Read Vec2Single from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        X = reader.ReadSingle()
        Y = reader.ReadSingle()
    }

File: generated-code/pascal/codecraft/TcpReadWrite/Vec2Single.pas
unit Vec2Single;

{$mode delphi}{$H+}

interface

uses
    Stream,
    SysUtils;

type
    // 2 dimensional vector.
    TVec2Single = class
        // `x` coordinate of the vector
        x: Single;
        // `y` coordinate of the vector
        y: Single;
        constructor Create(x: Single; y: Single);
        // Read Vec2Single from input stream
        class function ReadFrom(stream: TStream): TVec2Single; static;
        // Write Vec2Single to output stream
        procedure WriteTo(stream: TStream);
        function ToString: ansistring; override;
    end;

implementation

constructor TVec2Single.Create(x: Single; y: Single);
begin
    self.x := x;
    self.y := y;
end;

class function TVec2Single.ReadFrom(stream: TStream): TVec2Single;
var x: Single;
var y: Single;
begin
    x := stream.ReadSingle;
    y := stream.ReadSingle;
    result := TVec2Single.Create(x, y);
end;

procedure TVec2Single.WriteTo(stream: TStream);
begin
    stream.WriteSingle(x);
    stream.WriteSingle(y);
end;

function TVec2Single.ToString: ansistring;
begin
    result := 'Vec2Single {';
    result += 'x=';
    result += FloatToStr(x);
    result += ', ';  
    result += 'y=';
    result += FloatToStr(y);
    result += '}';
end;

end.